packages/headless-react/src/hooks/controller-hook.ts

```diff
@@ -0,0 +1,159 @@
+import type {Controller, CoreEngine, Unsubscribe} from '@coveo/headless';
+import {deepEqual} from 'fast-equals';
+import {useEffect, useReducer, DependencyList, useRef} from 'react';
+
+function useDeepCompareMemo<T>(
+  factory: () => T,
+  deepDeps: object,
+  shallowDeps?: DependencyList
+): T {
+  const lastState = useRef<{
+    value: T;
+    shallowDeps: DependencyList | undefined;
+    deepDeps: object;
+  } | null>(null);
+  const updateState = () => {
+    lastState.current = {value: factory(), shallowDeps, deepDeps};
+    return lastState.current.value;
+  };
+
+  if (!lastState.current) {
+    return updateState();
+  }
+  if (
+    shallowDeps?.length !== lastState.current.shallowDeps?.length ||
+    shallowDeps?.some(
+      (dep, i) => !Object.is(dep, lastState.current?.shallowDeps?.[i])
+    )
+  ) {
+    return updateState();
+  }
+  if (!deepEqual(deepDeps, lastState.current.deepDeps)) {
+    return updateState();
+  }
+  return lastState.current.value;
+}
+
+function useHasChanged(deps: DependencyList) {
+  const ref = useRef<DependencyList | null>(null);
+  if (ref.current === null) {
+    ref.current = deps;
+    return false;
+  }
+  if (
+    ref.current.length === deps.length &&
+    !deps.some((dep, i) => !Object.is(ref.current![i], dep))
+  ) {
+    return false;
+  }
+  ref.current = deps;
+  return true;
+}
+
+export type ReactController<T extends Controller> = Omit<
+  T,
+  'subscribe' | 'state'
+>;
+
+export interface UseControllerReturnType<T extends Controller> {
+  state: T['state'];
+  controller: ReactController<T>;
+}
+
+function bindMethods<T extends object>(methods: T, obj: object): T {
+  Object.entries(methods).forEach(([key, value]) => {
+    if (typeof value === 'function') {
+      (methods as Record<string, unknown>)[key] = value.bind(obj);
+    } else if (typeof value === 'object') {
+      bindMethods(value, (obj as Record<string, object>)[key]);
+    }
+  });
+  return methods;
+}
+
+function getOnStateChanged(subscribe: (listener: () => void) => Unsubscribe) {
+  return (listener: () => void) => {
+    let firstCall = true;
+    return subscribe(() => {
+      if (firstCall) {
+        firstCall = false;
+        return;
+      }
+      listener();
+    });
+  };
+}
+
+function buildController(
+  builder: (engine: CoreEngine, props?: object) => Controller,
+  engine: CoreEngine,
+  props?: object
+) {
+  const headlessController = builder(engine, props);
+  const {subscribe: _, state: __, ...methods} = headlessController;
+  return {
+    controller: bindMethods(methods, headlessController),
```

Comment:
Because some of our controllers reference `this`, and by default, `this` is whichever object is present before the period (`.`), which in this case would only be methods.